docs/src/tutorial.md

```diff
@@ -0,0 +1,211 @@
+
+After the program has been made as type stable as possible, it's time to use `@report_call` to find type errors.
+
+## Analyse methods with `@report_call`
+The function/macro pair [`report_call`](@ref) and [`@report_call`](@ref) works just like `(@)report_opt` - but where the latter reports dynamic dispatch, the former finds type errors.
+
+The `@report_call` macro analyses function calls like so:
+```@repl opt1
+@report_call sum(['a'])
+```
+
+In this example, JET found two possible type errors:
+* If the input vector is empty, the function call will error with a `MethodError` after attempting to call `zero(Char)`.
+* If the input vector has two or more elements, the call will error after attempting to call `+(::Char, ::Char)`.
+Note that these type errors show up even though the input vector had exactly one element, and so neither of these errors would actually occur at runtime if `sum(['a'])` had been executed.
+This happens because JET analyses the code on a _type_ level, only looking at the code generated with the input _types_. It does not analyze what will actually happen with the given input value.
+
+Note also that the two possible errors shown are mutally exclusive - no input will lead to both errors.
+Nonetheless, JET is able to detect both possibilities, because it analyses all possible branches in the generated function call.
+
+In contrast, if we analyse the same `sum` method on a `Vector{Int}` instead of `Vector{Char}`:
+```@repl opt1
+@report_call sum([1])
+```
+
+The two errors above do not appear. These errors do not apply to `Vector{Int}`, because `zero(Int)` is well-defined and so is `+(::Int, ::Int)`.
+This illustrates that JET does not analyze _methods_, but rather _function calls with specific types_. More precisely, JET analyses so-called _methodinstances_, since methodinstances are compiled whereas methods are not.
+
+## Analyse whole packages with `report_package`
+As shown above, _methods_ cannot be analyzed, but only _methodinstances_, i.e. methods plus the types of their arguments.
+Most packages, however, define only methods, and do not contain callsites. That is, they do not have any information about the types that these methods will be called with.
+
+However, JET is able to do limited analysis using only the method signature extracted from the method definition.
+For example, if I define this simple function:
+
+```@repl
+first_plus_n(itr, n::Real) = first(itr) + n;
+```
+
+, then _at the very least_, we can guarantee that `itr isa Any` and `n isa Real`.
+Hence, JET can analyze the methodinstance `first_plus_n(::Any, ::Real)`, using only the method definition.
+
+The JET function [`report_package`](@ref) extracts all method definitions in a package, and using the extracted signatures, runs `report_call` on them.
+For example, the package `BioSymbols` can be analysed like this:
+
+```
+julia> using JET
+
+julia> report_package(JET)
+[ output elided ]
```

Comment:
Well, I miss the previous example usage of `report_package` on `BioSymbols` package.

Reply:
Okay, let me see if I can figure out how to re-add it!

Reply:
If it's gonna be hard, no worries. We can refine this section in another day.

Reply:
Hm, it gives:
```julia
julia> using JET, BioSymbols

julia> report_package(BioSymbols)
[toplevel-info] virtualized the context of Main (took 0.033 sec)
[toplevel-info] entered into /home/jakob/.julia/packages/BioSymbols/M10HM/src/BioSymbols.jl
[toplevel-info] entered into /home/jakob/.julia/packages/BioSymbols/M10HM/src/nucleicacid.jl
[toplevel-info]  exited from /home/jakob/.julia/packages/BioSymbols/M10HM/src/nucleicacid.jl (took 0.28 sec)
[toplevel-info] entered into /home/jakob/.julia/packages/BioSymbols/M10HM/src/aminoacid.jl
[toplevel-info]  exited from /home/jakob/.julia/packages/BioSymbols/M10HM/src/aminoacid.jl (took 0.188 sec)
[toplevel-info] entered into /home/jakob/.julia/packages/BioSymbols/M10HM/src/workload.jl
[toplevel-info]  exited from /home/jakob/.julia/packages/BioSymbols/M10HM/src/workload.jl (took 0.282 sec)
[toplevel-info]  exited from /home/jakob/.julia/packages/BioSymbols/M10HM/src/BioSymbols.jl (took 3.64 sec)
═════ 2 toplevel errors found ═════
┌ @ /home/jakob/.julia/packages/BioSymbols/M10HM/src/workload.jl:1
│ ArgumentError: Package SnoopPrecompile not found in current path.
│ - Run `import Pkg; Pkg.add("SnoopPrecompile")` to install the SnoopPrecompile package.
│ Stacktrace:
│  [1] macro expansion
│    @ ./loading.jl:1163 [inlined]
│  [2] macro expansion
│    @ ./lock.jl:223 [inlined]
│  [3] require(into::Module, mod::Symbol)
│    @ Base ./loading.jl:1144
```
Despite having added `BioSymbols` to the JET docs Project, and SnoopPrecompile being a dep of BioSymbols. Not sure how to deal with that.

Reply:
Maybe we you need to `using Pkg; Pkg.activate(pkgdir(BioSymbols))` before doing `report_package`.
I think `report_package` should do this automatically though.

Reply:
Yes, that works! Okay, I'll leave out of the docs for now because it doesn't yet work, but I'll re-add it when it works :)

Reply:
Sounds great, I will let you know when I fix this problem.